pennylane/devices/qubit/sampling.py

```diff
@@ -183,6 +266,40 @@ def _measure_classical_shadow(
     return mp.process_state_with_shots(state, wires, shots.total_shots, rng=rng)
 
 
+def _measure_hamiltonian_with_samples(
+    mp: SampleMeasurement, state: np.ndarray, shots: Shots, is_state_batched: bool = False, rng=None
+):
+    # if the measurement process involves a Hamiltonian, measure each
+    # of the terms separately and sum
+    def _sum_for_single_shot(s):
+        results = measure_with_samples(
+            [ExpectationMP(t) for t in mp.obs.terms()[1]],
+            state,
+            s,
+            is_state_batched=is_state_batched,
+            rng=rng,
+        )
+        return sum(c * res for c, res in zip(mp.obs.terms()[0], results))
```

Comment:
In a subsequent PR, we could use `grouping_indices`, but I think this is fine for now.